### PR TITLE
Fetch Arrow - Streaming

### PR DESCRIPTION
**Pull request by pdet**
This PR Implements methods to stream query results into Arrow Objects in both Python and R lands.

### In Python:
```python
fetch_arrow_chunk(vectors_per_chunk=1,return_table=False)
```
* The Vectors Per Chunk parameter is an integer representing the number of vectors returned per fetch call (e.g., vectors_per_chunk = 2, would return up to 2048 elements since DuckDB's current default vector size is 1024).
* The Return Table parameter is a boolean that indicates if we should return an Arrow Table (if true) or a List of Record Batches (if false).

Example of usage:
```python 
import duckdb
duckdb_cursor = duckdb.connect().cursor()
duckdb_cursor.execute("CREATE table t as select range a from range(10000);")
query = duckdb_cursor.execute("SELECT a FROM t")
        
# Return 2 vectors and an Arrow Table
cur_chunk = query.fetch_arrow_chunk(2,True)
```

### In R:
```r
duckdb_fetch_arrow <- function(res,stream=FALSE,vector_per_chunk=1,return_table=FALSE)
 ```
* res  = Query result to be converted to an Arrow Object
* stream  = If we are streaming the query result or returning it all at once
* vector_per_chunk =  If streaming, how many vectors per chunk we should emit (Similar to Python)
* return_table =  If we return results as a list of RecordBatches or an Arrow Table (Similar to Python)

Example of usage:
```r 
con <- dbConnect(duckdb::duckdb())
dbExecute(con, paste0("CREATE table t as select range a from range(10000);"))
res <- dbSendQuery(con, "SELECT * FROM t", arrow=TRUE)

# Will stream the result into Arrow Tables with size up to 2* the vector size
cur_chunk <- duckdb::duckdb_fetch_arrow(res, TRUE,2,TRUE)
```

cc @jonkeane

### CODE REVIEW

**Comment by pdet** [comment]
@hannesmuehleisen Can you review the R side of things?

**Comment by hannesmuehleisen** [comment]
@pdet aye will do

File: tools/rpkg/src/statement.cpp [hunk 1]
```diff
@@ -449,14 +449,42 @@ struct AppendableRList {
...
+                              SEXP return_tablesexp) {
 	RProtector r;
+	RQueryResult *query_result_holder = (RQueryResult *)R_ExternalPtrAddr(query_resultsexp);
+	auto result = query_result_holder->result.get();
 	// somewhat dark magic below
 	SEXP arrow_name_sexp = r.Protect(RApi::StringsToSexp({"arrow"}));
 	SEXP arrow_namespace_call = r.Protect(Rf_lang2(Rf_install("getNamespace"), arrow_name_sexp));
 	SEXP arrow_namespace = r.Protect(RApi::REvalRerror(arrow_namespace_call, R_GlobalEnv));
-
+	bool stream = LOGICAL_POINTER(streamsexp)[0] != 0;
```
Comment (hannesmuehleisen) [inline tools/rpkg/src/statement.cpp:485]: perhaps add some checks here, `streamsexp` and `vector_per_chunksexp` need to be a LOGICAL of length 1, `vector_per_chunksexp` needs to be numeric of length 1. 
e.g.
````C++
if (TYPEOF(streamsexp) != LGLSXP || LENGTH(streamsexp) != 1) {
  Rf_error("stream parameter needs to be single-value logical");
}
````

**Comment by jonkeane** [comment]
This is very cool and glad to see it working. I wonder though if it would make sense to use [ExportRecordBatchReader](https://arrow.apache.org/docs/cpp/api/c_abi.html#c-stream-interface) instead of a list of record batches in this case so that the record batches don’t need to be materialized at all until they are being read by whatever they are being passed on to?

**Comment by hannesmuehleisen** [comment]
@jonkeane @pdet indeed I wonder whether we can directly return the reader object

**Comment by jonkeane** [comment]
Yes, exactly. You can “just” return the reader (or the pointer to the reader) and then whatever is picking it up can use `ImportRrcordBatchReader` and be good to go. 

(Also sorry for my brevity or typos – I’m on my phone right now!)

**Comment by pdet** [comment]
Hmm, the ExportRecordBatchReader function is marked with (visibility("default")) so I guess I can link to that at runtime, no problem.
But it is not clear to me how to construct the reader,  it seems that I would need a RecordBatchVector, and through skimming the arrow code it looks like this is a vector of all record batches. At that point, I would be materializing all batches before exporting it, no?

**Comment by pdet** [comment]
Oh wait, I guess  we can  inherit and overwrite the next method of the RecordBatchReader at run time because that's exported as well

**Comment by pdet** [comment]
So @jonkeane, I gave some thought about this over the weekend. I think the dynamically linking at runtime, although possible, is a bit hacky.
I'm wondering if it would be possible to extend Arrow's stream interface to allow us to produce these RecordBatchReaders similarly to our current and ArrowArray and ArrowArrayStream production.
Or even if there is another way of creating those that I'm unaware of.

**Comment by jonkeane** [comment]
@bkietz might have a better idea than I for how to extend this or if it exists already.

**Comment by bkietz** [comment]
@pdet I'm not sure what you're proposing. It sounds like you'd prefer to reimplement ExportRecordBatchReader and ImportRecordBatchReader rather than link to libarrow's

**Comment by pdet** [comment]
Hey @jonkeane 
Finally figured out that the interface I wanted was already there, so I managed to create the RecordBatchReaders without any linking :-).

To use it:

in R land :
```R
library("DBI")
dbExecute(con, paste0("CREATE table t as select range a from range(3000);"))
res <- dbSendQuery(con, "SELECT * FROM t", arrow=TRUE)
record_batch_reader <- duckdb::duckdb_fetch_record_batch(res)
```

in Python land:
```python
conn = duckdb.connect()
conn.execute("CREATE table t as select range a from range(3000);")
query = conn.execute("SELECT a FROM t")
record_batch_reader = query.fetch_record_batch()
```

**Comment by jonkeane** [comment]
Awesome, that looks great. I'll try and set some time aside today to pull this branch and try it out with some of our datasets to see how it works.

**Pull request merged**